### ISSUE TITLE
Glide64 For PJ64: 2.0.0.3 - black screen error.

### ISSUE THREAD

**Issue by Frank-74** [project64/project64#152]
Every time I close Glide's settings, I have a black screen.  So I can't adjust settings after starting a game.

Music still plays, and the game is still running, but just black screen.

**Comment by cxd4** [comment]
There is a memory leak viewable through the debug build of the plugin, after closing DllConfig/DllAbout.

Maybe fixing that would have something to do with it, but I have no understanding of C++ memory allocation whatsoever so can't have much of a further look to it.

**Comment by Frank-74** [comment]
A quick save with F5 before changing settings and restore.with F7 after gets the display back.  I'll use that for now.

**Comment by cxd4** [comment]
What happens for me is I open the config right after booting up Mario, click either OK/Cancel/X, and in either case, the screen quickly flashes black for a frame or more, then resumes updating the screen.

So maybe it is "supposed" to be black after updating the config/exiting it, but also it's supposed to restore the context state or flush updated commands to OpenGL but for some reason is only doing that on my system and not yours?

Anyone else able to reproduce this?

**Comment by cxd4** [comment]
Also does this issue happen only with Release mode or with Debug mode as well?
If it happens in Debug mode I can add a tracer for internal OpenGL state errors.

**Comment by Frank-74** [comment]
It happens in both Release and Debug builds.  Though I think this issue is less important than the memory leak, or failure to close glide properly when emulation ends.

At the moment, Glide crashes Project 64 when loading Glide, switching to Jabo and trying Glide again.  Maybe its a related issue.

**Comment by cxd4** [comment]
Some commits were just merged to affect the verbosity of the Debug build of this plugin.
It should now:
* track driver faults that lose the OpenGL context abruptly
* track any other OpenGL-related run-time errors in the plugin
* constantly flush the GL command queue before continuing with the currently running thread

Does any of this modify how this issue you are having behaves with the Debug build of PJGlide64 now?

**Comment by Frank-74** [comment]
I merged your pull before zilmar merged and missed one commit.  Think I've sorted out the mess that GitHub has gotten into and will build the new debug and release builds.  Maybe easier if I just delete everything and start again.  Though I have a branch I want to keep.  Damn I wish I could use GitHub for Windows (doesn't work on XP anymore).  Have to use the website all the time, and thats locking up and crashing Firefox all the time too.

**Comment by cxd4** [comment]
That last commit was fail on my part because I copied the line out of another plugin I did, then did Ctrl+S to save a last-minute fix for compiling on Linux not Windows, then forgot to hit Ctrl+S a second time after realizing I wasn't done yet.  So the last two commits in my pull request shouldn't be very relevant.

> Damn I wish I could use GitHub for Windows (doesn't work on XP anymore). Have to use the website all the time, and thats locking up and crashing Firefox all the time too.

You use GitHub website to download the entire repository and extract from a zip?
I did that often sometimes back before I was familiar with Git console.

I think if you just do `git checkout master` then `git pull https://github.com/project64/project64.git`, it should pull all updates zilmar did to the master branch to your own master branch of Frank-74/project64.git.  That way you wouldn't have to download anything using the GitHub site to keep the repo.

**Comment by Frank-74** [comment]
My master branch has commits from deathdroid mempak changes, and stops me from merging because of conflicts.  My Clean branch I created from a patch I did earlier is currently the same as project64-master, even though its 5 commits ahead of the pjmaster branch.  Only I can't do any pull requests from the Clean branch, because it tries to revert the 5 outstanding commits.

I need to delete I think.

How do I get the GitHub console?  I cannot install GitHub on XP.

**Comment by cxd4** [comment]
When I am talking about the console, I'm talking really just about Git itself.  Git is the version control system, and GitHub is one of many websites that gives us a graphical frontend/shortcut to do some things over HTTP instead.  If you have Git GUI installed to your computer, then somewhere in your start menu/installed program entries should be "Git Bash".  That's the bash Unix terminal for directly executing Git commands.

If you merged death droid's mempak changes to your own master, then I think you are right.  I'm not very experienced with Git myself, but so far I have had to delete my repository off my GitHub account, then go to github.com/project64, click the "Fork" button to re-fork zilmar's instance of Project64 to my repository again to start over.  Then, `git clone https://github.com/cxd4/project64.git`.

**Comment by Frank-74** [comment]
I cannot install anything from git, I get the message, "The easiest way to use GitHub on Windows
is unfortunately not supported by your OS. Upgrade to Windows 7 or newer.".

So I cannot get the console/bash Unix terminal.  Is probably why I think I'm the only XP user here.

**Comment by cxd4** [comment]
And you went here to install Git?  http://git-scm.com/download/win
If so, then I guess the terminal version really doesn't work on XP, which I find to be pretty f*cked.

Maybe you would have to download it through the web browser then if you can't execute Git commands on your own machine.  I'm surprised it's even possible to do multiple pull requests to zilmar just using a GitHub web interface.  I suppose if you were familiar with IRC you could join us on #project64 at Freenode if there was anything else to troubleshoot about it.

**Comment by Frank-74** [comment]
Just downloaded OpenGL Extensions Viewer to test my system.  It passed all tests according to the program, but I only saw a flash of each test for a fraction of a second and then black screen.  Very weird.  So I just installed ATI Catalyst Drivers from AMD website, same version as I had before, but my old ones were from motherboard manufacturer Packard Bell.  And now Glide isn't staying on a black screen.

By the way, found out my integrated Radeon Xpress 200 supports only upto OpenGL 2.1, partial support for 3.0 and 3.1.  But OGL Extensions Viewer tests are still not displaying properly, 1.5 test crashes the Extensions Viewer program, and its finding other errors about missing entries.  I've re-installed everything, at least Glide is no longer staying on a black screen, and no errors in display yet.

So to sum up:
Glide needs me to use ATI drivers from AMD, but OpenGL tests fail.
Packard Bell motherboard drivers pass OpenGL tests but Glide fails.

**Comment by cxd4** [comment]
lol.  No wonder none of the rest of us could reproduce this.
The fact that it was a driver issue isn't surprising, but to the degree that you mentioned...that is surprising.

Sometimes drivers have some totally weird bugs.  There used to be arbitrary limits, like you couldn't have X and Y enabled/disabled at the same time while Z is enabled.  Usually these rules are supposed to be abolished, but sometimes GPU drivers can have weird bugs in them.  Maybe after DllConfig temporarily killed the context to update the state config then recreated it, it enabled some blender/lightning/texture/etc. modes that your own personal driver didn't like and had a problem conforming to the GL spec about it.

I know on my NVIDIA GPU I had to constantly make sure GL_TEXTURE_2D was glDisable'd if I ever wanted to test glDrawPixels, or it would totally freeze the rendering thread to a sluggish state.  These drivers just have weird bugs in them sometimes.

But there is no change from using the Debug build of PJGlide64 from the recent changes I did?  No unexpected context loss or "invalid operation" error messages...

**Comment by Frank-74** [comment]
I've deleted everything and am rebuilding the solution from project64-master.zip
I'll run the debug build as soon as its done.  May have to re-install the Packard Bell drivers too.  Will take a little while...

**Comment by Frank-74** [comment]
[quote]But there is no change from using the Debug build of PJGlide64 from the recent changes I did? No unexpected context loss or "invalid operation" error messages...[/quote]
Just tested, and no error messages at all.  The only thing I got was about the memory leak when closing Project64, and the leak csv log.

**Comment by V1del** [comment]
It really isn't that surprising windows update driver or laptop vendor drivers tend to give no shit what opengl is or does, I really don't know how they can fuck this up so badly all the time, when the drivers from the vendor provide proper support. Are they intentionally patching it out or what?

**Comment by cxd4** [comment]
I wrote a basic OpenGL tester once for WGL/Windows, but all it really did was analyze the full GL_EXTENSIONS list.  I really don't have the kind of modern familiarity to OpenGL to actually test running that stuff in practice.  So maybe something has to be glDisable'd on Frank-74's end because his driver as with so many other cases doesn't like rendering to texture/screen with a certain mode enabled.

There's only so much I can suggest to change about the GL code though since I don't have his system and can't test.  I can try having a closer look at how DllConfig is halting/resuming the context for the config when I get back home.

**Comment by cxd4** [comment]
I would say, try adding PFD_SUPPORT_GDI to the WGL initialization code for the context creation to Glitch64's usage of GL, **or**, try "run in 256 colors" compatibility mode on Project64.exe windows compat. settings to force using Microsoft's software implementation of OpenGL instead of your AMD drivers, to see if you still have the bug then.  However, I'm not really sure that Glitch64 would support OpenGL 1.1...so I imagine this wouldn't work.  Heh, wonder if it's possible for me to add an OpenGL 1.1 fallback in Glitch64 for really old machines.

**Comment by cxd4** [comment]
@Frank-74 , I know this is going to be painfully slow, but try using Mesa3D.
I uploaded my old build of Mesa3D 7.5.1:  https://dl.dropboxusercontent.com/u/16494013/mesa_w32.zip

Extract `mesa/opengl32.dll` to the root directory of Project64 emulator that has `Project64.exe` in it.
This will cause Windows to load Project64 using the Mesa3D software implementation of OpenGL instead of Microsoft's own software implementation of it, without the proprietary video drivers AMD ships with your card for OpenGL.

Again this is going to be painfully slow since this will cause Glide64 to operate in pure software, but does it fix your black screen on config issue?

**Comment by Frank-74** [comment]
I fixed my black screen issue, by completely removing everything ATI and reverting to older Packard Bell drivers, the drivers that originally came with my PC.  Radeon Xpress 200 ver 8.1 by Packard Bell.  They're ancient drivers, but seem to work well.  Apart from Software depth buffer causes a black screen, disabling it brings my display back.

Tried the mesa3d opengl anyway, and I get 1 or 2 VI's.

**Comment by cxd4** [comment]
Mesa3D is 1 or 2 VI/s for me as well.  I was hoping it would be enough to open the config, then click OK/close it to confirm that the black screen issue is fixed.  Although, I'm 95% confident that it should have fixed it for you.

Yeah, I saw that you were able to pick a different poison with the other drivers to fix the black screen issue.  I don't mind this issue staying open though because then I can try to fix it for people who have the same drivers.

**Comment by Frank-74** [comment]
Ok, and anyway, when I did have the problem, the official Glide Final never had the problem.  So it is related to only the PJ_Glide version.

**Comment by cxd4** [comment]
okay a few things frank,

first, re: no Git on Windows XP,
Here is a list of older versions of Git Bash console--
https://www.kernel.org/pub/software/scm/git/?C=M;O=A
It goes all the way back to 2005 (and skips miraculously to 2013, sadly), so hopefully something older can install on your WinXP so you don't have to update the repository off HTTP through FireFox all the time.

second, more debugging commits for Glide64 #ifdef _DEBUG were merged recently

Third, what if you try this:
```c
void CALL UpdateScreen (void)
{
    newSwapBuffers ();
```
up at the very top of the `UpdateScreen` function in `main.cpp`?

**Comment by vgturtle127** [comment]
@Frank-74 Have you tried TortoiseGit? I think older versions supported XP...

**Comment by oddMLan** [comment]
@Frank-74 Is this still an issue?

Removing WxWidgets might have helped this case. That, and the somewhat recent merge of Glide64mk2 improvements.
If it didn't, we'll have to find the commit that caused the regression.

Out of curiosity, does the same thing happen if you check "Software depth rendering"?
@AmbientMalice disabled it globally and enabled it only for the games that actually need it like Resident Evil 2. It might be triggering something in your old IGP.

**Comment by Frank-74** [comment]
No longer an issue with configure graphics.

But now get black screen when running a second game with Glide.  If I End Emulation and restart the same game (or different game with PJ64Glide) I get black screen.

**Comment by Frank-74** [comment]
I'm sure this is a driver issue.  It happens with GLideN64, Glide64Final and PJ64Glide.

Black screen on closing/opening any plugin using OpenGL.

I've just updated my Intel HD Graphics driver from Hewlett Packard.  No difference.
Now I'll try Intel's drivers.............

**Comment by Frank-74** [comment]
Fixed :)

Latest Intel HD Graphics driver using the Intel driver update utility to autodetect my hardware.

**Comment by topkex** [comment]
My toaster has Intel GMA, the drivers will never be updated. Is there a workaround for it?

**Comment by Frank-74** [comment]
Go here and download the latest driver for your chipset.
https://downloadcenter.intel.com/product/80939/Graphics-Drivers

Click on Additional selections available.  Scroll down for Intel® Graphics Media Accelerator.  Choose your driver, download and install.

Even if they were last updated in 2013 or something, they may fix your black screen.
My laptop had drivers from HP.  I had to remove them first from Device Manager, reboot, wait for windows to detect the new hardware and install a default Intel driver, then install the new driver.

**Comment by topkex** [comment]
Yeah I already have the latest driver it's from November 2012.
For sure it has something to do with opengl, angrylions RDP has the same problem.
Intel's OpenGL implementation has been shady for some time, Oh well.

**Comment by cxd4** [comment]
Intel have been known for their flaky GL drivers.  It's that network effect from the proprietary Direct3D.

If your vendor (Intel or otherwise) for your GPU and operating system doesn't provide a bug fix to these proprietary drivers, then the only thing you really can do (besides avoiding that vendor and replacing your hardware) is install a free driver.  A leading project for this is Mesa3D:  If you can get it to compile for Windows, great, but I never figured out how.

If you're experienced with make and build systems (not coding, not programming...), go to Mesa's site and build the full driver with their Wayland accelerator compiled in for 3-D performance.  I wasted my time trying to figure out how to do this, personally.  Oh, or you could just wipe Windows and use a free-software operating system, which comes with Mesa anyway.

**Comment by Frank-74** [comment]
With every PC, notebook, laptop I have.  The manufacturer drivers (vendor) have had this black screen problem with integrated gfx chipsets.

The only solution is to install reference drivers from Intel/AMD/NVidia.

The driver version may be the same as your vendor drivers, but they're more complete.  And work as they should.

**Comment by LegendOfDragoon** [comment]
I like how cxd4's code has the ability to work around the black screen issue, by doing the initialization in InitiateGfx.

**Comment by LegendOfDragoon** [comment]
@Frank-74 are you able / willing to test with the drivers you originally had? I'm curious if it's still an issue.

**Comment by Frank-74** [comment]
@LegendOfDragoon I'm not willing to go back to a broken state.  The problem is with gfx drivers.  I've had the problem with my old PC, a netbook and my current laptop.  The solution is to install the reference drivers from intel, if you have intel.  The vendor drivers for my HP laptop are useless.

Closing this issue now, as its not any fault of PJ64, only fault is GL support ID by the driver.

**Comment by project64** [comment]
Also hopefully will improve with the changes for angle as well .. just starting to look at cleaning up glide64 code .. it is a real mess..

**Comment by vgturtle127** [comment]
> it is a real mess...
@project64 You got that right. 😆
